IntegrationTest/ExerciseEndpointsTest.cs

```diff
@@ -0,0 +1,576 @@
+
+        var userId = 1;
+        var roles = new List<Roles> { Roles.Instructor };
+        _client.AddRoleAuth(userId, roles);
+        var requestBody = CreateExerciseRequestBody();
+
+        var response = await _client.PutAsync("v1/exercises/1", requestBody);
+
+        Assert.Equal(HttpStatusCode.InternalServerError, response.StatusCode);
+    }
+
+    [Fact]
+    public async Task UpdateExercise_FailedToValidateSolution_ShouldReturn_400()
+    {
+        using var scope = _factory.Services.CreateScope();
+        var haskellServiceSub = scope.ServiceProvider.GetService<IHaskellService>();
+        var exerciseRepoSub = scope.ServiceProvider.GetService<IExerciseRepository>();
+
+        exerciseRepoSub.VerifyExerciseAuthorAsync(Arg.Any<int>(), Arg.Any<int>()).Returns(true);
+        var solutionRunnerResponse = new SolutionRunnerResponse
+        {
+            Action = ResponseCode.Failure,
+            ResponseDto = new SolutionResponseDto("failure", null, new List<SolutionTestcaseResultDto> { new SolutionTestcaseResultDto(5, "test", null, null) })
+        };
+        haskellServiceSub.SubmitSubmission(Arg.Any<SubmissionDto>()).Returns(solutionRunnerResponse);
+        exerciseRepoSub.UpdateExerciseAsync(Arg.Any<ExerciseDto>(), Arg.Any<int>()).Returns(Result.Ok());
+
+        var userId = 1;
+        var roles = new List<Roles> { Roles.Instructor };
+        _client.AddRoleAuth(userId, roles);
+        var requestBody = CreateExerciseRequestBody();
+
+        var response = await _client.PutAsync("v1/exercises/1", requestBody);
+
+        Assert.Equal(HttpStatusCode.BadRequest, response.StatusCode);
+    }
+
+    [Fact]
+    public async Task UpdateExercise_CompilationErrorOnSolutionRunner_ShouldReturn_400()
+    {
+        using var scope = _factory.Services.CreateScope();
+        var haskellServiceSub = scope.ServiceProvider.GetService<IHaskellService>();
+        var exerciseRepoSub = scope.ServiceProvider.GetService<IExerciseRepository>();
+
+        exerciseRepoSub.VerifyExerciseAuthorAsync(Arg.Any<int>(), Arg.Any<int>()).Returns(true);
+        var solutionRunnerResponse = new SolutionRunnerResponse
+        {
+            Action = ResponseCode.Error,
+            ResponseDto = new SolutionResponseDto("error", "compilation error", null)
+        };
+        haskellServiceSub.SubmitSubmission(Arg.Any<SubmissionDto>()).Returns(solutionRunnerResponse);
+        exerciseRepoSub.UpdateExerciseAsync(Arg.Any<ExerciseDto>(), Arg.Any<int>()).Returns(Result.Ok());
+
+        var userId = 1;
+        var roles = new List<Roles> { Roles.Instructor };
+        _client.AddRoleAuth(userId, roles);
+        var requestBody = CreateExerciseRequestBody();
+
+        var response = await _client.PutAsync("v1/exercises/1", requestBody);
+
+        Assert.Equal(HttpStatusCode.BadRequest, response.StatusCode);
+    }
+
+    [Fact]
+    public async Task UpdateExercise_NoAuthentication_ShouldReturn_401()
```

Comment:
same idea here you can probably delete everything except the response and assert